### PR TITLE
Add independent backtesting to run backtesting within an existing bot 

### PR DESCRIPTION
**Pull request by GuillaumeDSM**
requires https://github.com/Drakkar-Software/OctoBot-Trading/pull/88
requirs https://github.com/Drakkar-Software/OctoBot-Evaluators/pull/28

### CODE REVIEW

File: octobot_backtesting/api/backtesting.py [hunk 1]
```diff
@@ -13,24 +13,19 @@
...
+from octobot_backtesting.independent_backtesting import IndependentBacktesting
 from octobot_commons.constants import CONFIG_ENABLED_OPTION
 
 from octobot_backtesting.backtesting import Backtesting
 from octobot_backtesting.constants import CONFIG_BACKTESTING, CONFIG_BACKTESTING_DATA_FILES
 
 
-def create_backtesting(config, data_files) -> Backtesting:
-    return Backtesting(config, data_files)
+def create_independent_backtesting(config, data_files) -> Backtesting:
```
Comment (Herklos) [inline octobot_backtesting/api/backtesting.py]: A backtesting should always be independent from the current bot isn't it ?

File: octobot_backtesting/backtesting.py [hunk 2]
```diff
@@ -35,19 +35,32 @@ def __init__(self, config, backtesting_files):
...
         except Exception as e:
             self.logger.error(f"Error when initializing backtesting : {e}.")
             self.logger.exception(e)
 
+    async def stop(self):
+        await self.delete_time_channel()
+        self.time_channel.time_updater = None
+        self.time_updater.backtesting = None
+
+    async def delete_time_channel(self):
```
Comment (Herklos) [inline octobot_backtesting/backtesting.py]: :+1:

File: octobot_backtesting/api/data_file.py [hunk 3]
```diff
@@ -15,26 +15,22 @@
...
-        print(e)
-        description = None
-    finally:
-        if database is not None:
-            await database.stop()
-    return description
+    description = await data_manager.get_file_description(path.join(data_path, file_name))
+    if description:
+        return {
+            DataFormatKeys.SYMBOLS.value: description[DataFormatKeys.SYMBOLS.value],
```
Comment (Herklos) [inline octobot_backtesting/api/data_file.py]: Why this code isn't in `data_manager.get_file_description` anymore ?

File: octobot_backtesting/independent_backtesting.py [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+            self.backtestings = []
+
+        except ImportError as e:
+            self._log_import_error()
+            raise e
+        except Exception as e:
+            self.logger.exception(e)
+            self.logger.error(f"Error when stopping independent backtesting: {e}")
+
+    def memory_leak_checkup(self, to_check_elements):
```
Comment (Herklos) [inline octobot_backtesting/independent_backtesting.py]: :+1:

File: octobot_backtesting/independent_backtesting.py [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+                # Warning: Python debugger can add references when watching an element
+                element = to_check_elements[i]
+                # Now expect 3 references because the above element variable adds a reference
+                self.logger.error(f"Too many remaining references on the {element.__class__.__name__} element after "
+                                  f"{self.__class__.__name__} run, the garbage collector won't free it "
+                                  f"(expected a maximum of 3 references): {getrefcount(element)} actual references")
+
+    # Use check_remaining_objects to check remaining objects from garbage collector after calling stop().
+    # Warning: can take a long time when a lot of objects exist
+    def check_remaining_objects(self):
```
Comment (Herklos) [inline octobot_backtesting/independent_backtesting.py]: :+1:

File: octobot_backtesting/independent_backtesting.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+from octobot_backtesting.enums import DataFormatKeys
+from octobot_commons.constants import CONFIG_ENABLED_OPTION, CONFIG_CRYPTO_CURRENCIES, CONFIG_CRYPTO_PAIRS
+from octobot_commons.enums import PriceIndexes
+from octobot_commons.errors import ConfigTradingError
+from octobot_commons.logging.logging_util import get_logger
+from octobot_commons.symbol_util import split_symbol
+from octobot_commons.time_frame_manager import find_min_time_frame
+
+
+class IndependentBacktesting:
```
Comment (Herklos) [inline octobot_backtesting/independent_backtesting.py]: Should we cythonize this class ?

File: octobot_backtesting/independent_backtesting.py [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+        self.backtesting_config = {
+            CONFIG_BACKTESTING: {},
+            CONFIG_CRYPTO_CURRENCIES: {},
+            CONFIG_EXCHANGES: {},
+            CONFIG_TRADER: {},
+            CONFIG_SIMULATOR: {},
+            CONFIG_TRADING: {},
+        }
+
+    async def get_dict_formatted_report(self):
```
Comment (Herklos) [inline octobot_backtesting/independent_backtesting.py]: A little bit too long method I think.

File: octobot_backtesting/api/backtesting.py (hunk 1 above)
Comment (GuillaumeDSM) [reply octobot_backtesting/api/backtesting.py]: Right, I think it's better to explicitly have it in the method's name not to confuse it with a backtesting mode bot run

File: octobot_backtesting/independent_backtesting.py [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+        from octobot_trading.constants import CONFIG_TRADER_RISK, CONFIG_TRADING, CONFIG_SIMULATOR, \
+            CONFIG_STARTING_PORTFOLIO, CONFIG_SIMULATOR_FEES,CONFIG_TRADER_REFERENCE_MARKET
+        self.backtesting_config[CONFIG_TRADING][CONFIG_TRADER_RISK] = self.risk
+        self.backtesting_config[CONFIG_TRADING][CONFIG_TRADER_REFERENCE_MARKET] = self._find_reference_market()
+        self.backtesting_config[CONFIG_SIMULATOR][CONFIG_STARTING_PORTFOLIO] = self.starting_portfolio
+        self.backtesting_config[CONFIG_SIMULATOR][CONFIG_SIMULATOR_FEES] = self.fees_config
+        self._add_config_default_backtesting_values()
+        self._add_crypto_currencies_config()
+
+    def _find_reference_market(self):
```
Comment (Herklos) [inline octobot_backtesting/independent_backtesting.py]: Is this a backtesting specific code ?

File: octobot_backtesting/independent_backtesting.py [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+
+        if market_begin and market_end and market_begin > 0:
+            market_delta = market_end / market_begin - 1 if market_end >= market_begin \
+                else 1 - market_begin / market_end
+        else:
+            market_delta = 0
+
+        return market_delta
+
+    async def _init_evaluators(self):
```
Comment (Herklos) [inline octobot_backtesting/independent_backtesting.py]: I think the following methods concerning the bot initialization with other module API should be in another dedicated class like `OctoBotBacktesting`.

File: octobot_backtesting/api/data_file.py (hunk 3 above)
Comment (GuillaumeDSM) [reply octobot_backtesting/api/data_file.py]: data_manager.get_file_description is generic, this one in is used to display info to the user about a specific data file

File: octobot_backtesting/independent_backtesting.py (hunk 8 above)
Comment (GuillaumeDSM) [reply octobot_backtesting/independent_backtesting.py]: Yes, in other cases we should use the config ref market

File: octobot_backtesting/independent_backtesting.pxd [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+#  Lesser General Public License for more details.
+#
+#  You should have received a copy of the GNU Lesser General Public
+#  License along with this library.
+
+
+from octobot_backtesting.octobot_backtesting cimport OctoBotBacktesting
+
+cdef class IndependentBacktesting:
```
Comment (Herklos) [inline octobot_backtesting/independent_backtesting.pxd]: :+1:

**Pull request merged**